File: LeetCode_347/main.cpp
/**
    LeetCode_347
    https://leetcode.com/problems/top-k-frequent-elements/description/
    MatrixDeity, 2018.
*/

#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        vector<int> ans;
        unordered_map<int, int> um;
        for (auto item : nums) {
            ++um[item];
        }
        vector<vector<int>> vec(nums.size() + 1);
        for (auto p : um) {
            vec[p.second].push_back(p.first);
        }
        for (int i = vec.size() - 1; i >= 0; --i) {
            for (auto item : vec[i]) {
                if (k-- == 0)
                    return ans;
                ans.push_back(item);
            }
        }
        return ans;
    }
};

int main() {
    return 0;
}

File: LeetCode_347/main.go
/**
  LeetCode_347
  https://leetcode.com/problems/top-k-frequent-elements/
  MatrixDeity, 2023.
*/

package leetcode347

import "sort"

type Pair struct {
	Val int
	Num int
}

func topKFrequent(nums []int, k int) []int {
	counter := map[int]int{}

	for _, num := range nums {
		counter[num]++
	}

	pairs := make([]Pair, 0, len(nums))
	for key, val := range counter {
		pairs = append(pairs, Pair{Val: key, Num: val})
	}

	sort.Slice(pairs, func(i, j int) bool {
		return pairs[i].Num > pairs[j].Num
	})

	res := make([]int, 0, k)
	for _, pair := range pairs[:k] {
		res = append(res, pair.Val)
	}

	return res
}
